### ISSUE TITLE
Almost successfully reproducing open-r1/OpenR1-Qwen-7B based on Qwen/Qwen2.5-Math-7B-Instruct. Here are the training configurations.

### ISSUE THREAD

**Issue by kiseliu** [huggingface/open-r1#545]
**After encountering several errors and incorrect results**, I'd like to share my experience reproducing `open-r1/OpenR1-Qwen-7B` based on `Qwen/Qwen2.5-Math-7B-Instruct`.

**The training commands below are configured for a node of 8 x H100s (80GB)**.

### 1. Modify the config file of Qwen/Qwen2.5-Math-7B-Instruct
 
After downloading the model `Qwen/Qwen2.5-Math-7B-Instruct`, we should modify `the model config file` following https://huggingface.co/open-r1/OpenR1-Qwen-7B/blob/main/config.json

### 2. Modify the training recipes correctly

If you **follow the official installation steps** and **run the following training command**:

```
accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/sft.py \
    --config recipes/OpenR1-Qwen-7B/sft/config.yaml
```

The first issue you’ll encounter is https://github.com/huggingface/open-r1/issues/366.

To resolve this issue, the effective way is to **modify the corresponding recipe**: https://github.com/huggingface/open-r1/blob/main/recipes/OpenR1-Qwen-7B/sft/config.yaml by **changing the line 29**
```
- use_liger_kernel: true
+ use_liger: true
```

Other solutions will lead to other issues, such as:
- Downgrading `trl` to `0.15.x`: it might not work or have this issue: https://huggingface.co/open-r1/OpenR1-Qwen-7B/discussions/7 ; 
- Set `use_liger_kernel: false`: then we cannot set the `gradient_accumulation_steps: 2` since **it will bring an OOM problem**;

### 3. Modify the sft.py file

Following this issue: https://github.com/huggingface/open-r1/pull/494, we could change the `sft.py` https://github.com/huggingface/open-r1/blob/main/src/open_r1/sft.py as follows:

```
- tokenizer.pad_token = tokenizer.eos_token
+ if tokenizer.pad_token is None:
+    tokenizer.pad_token = tokenizer.eos_token
```

### 4. The reproducing results on AIME24, MATH-500

I tested the saved model (the total step is 3219 steps), and the performance is:
- **AIME24: 46.7**;
- **MATH-500: 92.4**;

I tested `open-r1/OpenR1-Qwen-7B`, and the performance is:
- **AIME24: 50.0**;
- **MATH-500: 92.8**;

It seems that this model is saved at step 3150.

**Comment by Tim-Siu** [comment]
I wonder if replacing `DataCollatorForLanguageModeling` with `DataCollatorForCompletionOnlyLM` would further boost the performance.

Rightnow, the OpenR1 recipes train on both prompts and response. It doesn't seem like common practices.

**Comment by kiseliu** [comment]
@Tim-Siu I am not sure if the performance will boost. 

For the current version, it only costs 12 hours for sft on one node of 8 x H100s (80GB).

If replacing `DataCollatorForLanguageModeling` with `DataCollatorForCompletionOnlyLM`, then we cannot set `packing=True`. I think the training time will take much longer like 20+ hours.

**Comment by VulDetect-llm** [comment]
how to run 7b model on multiple gpus, 1 A800 encounters oom errors

**Comment by kiseliu** [comment]
> Thank you so much!! I had two questions:
> 
> 1. In terms of efficiency didn't you change other parameters? Because doing exactly how you say I get expected time of run of 60 hours on 8 A100 of 80GB but should be ~x2.5 your reported time of 12hours (since you use H100s) so still quite off. I guess you must have changed the attn_implementation to flash_attention_2 right? Also as a side note, you can increase the per_device_train_batch_size to 2 for 8 Gpus of 80GB (don't forget to lower the grad_accum to 1) and still works without OOM. Then I get ~23hrs of expected runtime which is more what one would expect from H100s to A100s.
> 2. Did you try replicating other models such as the Qwen2.5-1.5B-Instruct? Because there I can't replicate.

@juliancodaforno 

1. I didn’t change any other parameters. The `attn_implementation` I used was the default value (`sdpa`).

   Yes, setting `per_device_train_batch_size=2` and `gradient_accumulation_steps=1` should work without causing OOM errors.

2. I previously attempted to run the code for Qwen2.5-1.5B-Instruct but did so incorrectly. I won’t retry it.

**Comment by XYaoooo** [comment]
hi, what is your global bz?

**Comment by kiseliu** [comment]
> hi, what is your global bz and learning rate? thx

@XYaoooo Hi, the global batch size is set to 16, the learning rate is 5.0e-05. All the hyperparameters are set to their default values. See the config file https://github.com/huggingface/open-r1/blob/main/recipes/OpenR1-Qwen-7B/sft/config.yaml .

**Comment by XYaoooo** [comment]
> > hi, what is your global bz and learning rate? thx
> 
> [@XYaoooo](https://github.com/XYaoooo) Hi, the global batch size is set to 16, the learning rate is 5.0e-05.

Hi, one more question, what the version of your trl and deepspeed? Thx

**Comment by HwangYej1** [comment]
Have you modified "max_position_embeddings": 4096 in the Qwen2.5-Math-7B-Instruct model file config?

**Comment by HwangYej1** [comment]
If you don't keep max_positions_embedding: 4096, how do you set the hyperparameter on the " MODEL_ARGS="pretrained=$MODEL,dtype=bfloat16,max_model_length=32768,gpu_memory_utilization=0.6,data_parallel_size=$NUM_GPUS,generation_parameters={max_new_tokens:2048,temperature:0.6,top_p:0.95}" " line when you test with lighteval?

**Comment by kiseliu** [comment]
> > > hi, what is your global bz and learning rate? thx
> > 
> > 
> > [@XYaoooo](https://github.com/XYaoooo) Hi, the global batch size is set to 16, the learning rate is 5.0e-05.
> 
> Hi, one more question, what is the version of your trl and deepspeed? Thx

trl==0.16.0; deepspeed=0.15.4;

**Comment by kiseliu** [comment]
> If you don't keep max_positions_embedding: 4096, how do you set the hyperparameter on the " MODEL_ARGS="pretrained=$MODEL,dtype=bfloat16,max_model_length=32768,gpu_memory_utilization=0.6,data_parallel_size=$NUM_GPUS,generation_parameters={max_new_tokens:2048,temperature:0.6,top_p:0.95}" " line when you test with lighteval?

max_model_length=32768, max_new_tokens=32768;

**Comment by XCF-Mike** [comment]
> **After encountering several errors and incorrect results**, I'd like to share my experience reproducing `open-r1/OpenR1-Qwen-7B` based on `Qwen/Qwen2.5-Math-7B-Instruct`.
> 
> **The training commands below are configured for a node of 8 x H100s (80GB)**.
> 
> ### 1. Modify the config file of Qwen/Qwen2.5-Math-7B-Instruct
> After downloading the model `Qwen/Qwen2.5-Math-7B-Instruct`, we should modify `the model config file` following https://huggingface.co/open-r1/OpenR1-Qwen-7B/blob/main/config.json
> 
> ### 2. Modify the training recipes correctly
> If you **follow the official installation steps** and **run the following training command**:
> 
> ```
> accelerate launch --config_file recipes/accelerate_configs/zero3.yaml src/open_r1/sft.py \
>     --config recipes/OpenR1-Qwen-7B/sft/config.yaml
> ```
> 
> The first issue you’ll encounter is [#366](https://github.com/huggingface/open-r1/issues/366).
> 
> To resolve this issue, the effective way is to **modify the corresponding recipe**: https://github.com/huggingface/open-r1/blob/main/recipes/OpenR1-Qwen-7B/sft/config.yaml by **changing the line 29**
> 
> ```
> - use_liger_kernel: true
> + use_liger: true
> ```
> 
> Other solutions will lead to other issues, such as:
> 
> * Downgrading `trl` to `0.15.x`: it might not work or have this issue: https://huggingface.co/open-r1/OpenR1-Qwen-7B/discussions/7 ;
> * Set `use_liger_kernel: false`: then we cannot set the `gradient_accumulation_steps: 2` since **it will bring an OOM problem**;
> 
> ### 3. Modify the sft.py file
> Following this issue: [#494](https://github.com/huggingface/open-r1/pull/494), we could change the `sft.py` https://github.com/huggingface/open-r1/blob/main/src/open_r1/sft.py as follows:
> 
> ```
> - tokenizer.pad_token = tokenizer.eos_token
> + if tokenizer.pad_token is None:
> +    tokenizer.pad_token = tokenizer.eos_token
> ```
> 
> ### 4. The reproducing results on AIME24, MATH-500
> I tested the saved model (the total step is 3219 steps), and the performance is:
> 
> * **AIME24: 46.7**;
> * **MATH-500: 92.4**;
> 
> I tested `open-r1/OpenR1-Qwen-7B`, and the performance is:
> 
> * **AIME24: 50.0**;
> * **MATH-500: 92.8**;
> 
> It seems that this model is saved at step 3150.

Hi, for the third 

`- tokenizer.pad_token = tokenizer.eos_token
+ if tokenizer.pad_token is None:
+    tokenizer.pad_token = tokenizer.eos_token`

I made the same changes, but the problem in #494 was not resolved.
For this issue, I also tried different versions of trl, including trl==0.15.1, trl==0.16.0, and trl==0.16.1, but none of them helped resolve the problem.
Are there any other changes that should be made?

**Comment by kiseliu** [comment]
@XCF-Mike hi, I think it might be that the code base have updated. Maybe you can check this issue #604 to give a try.

**Comment by StarLooo** [comment]
> Thank you so much!! I had two questions:
> 
> 1. In terms of efficiency didn't you change other parameters? Because doing exactly how you say I get expected time of run of 60 hours on 8 A100 of 80GB but should be ~x2.5 your reported time of 12hours (since you use H100s) so still quite off. I guess you must have changed the attn_implementation to flash_attention_2 right? Also as a side note, you can increase the per_device_train_batch_size to 2 for 8 Gpus of 80GB (don't forget to lower the grad_accum to 1) and still works without OOM. Then I get ~23hrs of expected runtime which is more what one would expect from H100s to A100s.
> 2. Did you try replicating other models such as the Qwen2.5-1.5B-Instruct? Because there I can't replicate.
Thanks for sharing!
I have also observed a similar phenomenon. Specifically, I used the same setup (per_device_train_bsz=1 & gradient_accumulation=2, attn_implementation using sdpa) and train the default split of open-r1-220K with 3 epochs on 8*H100 GPUs. After packing, the total number of train steps was 3222. However, the tqdm bar showed that it took 14 hours to complete just approximately 1200 training steps.
This is quite slow than the reported 12hours for total training time cost.

**Comment by StarLooo** [comment]
I got the following performance using this train recipe:
The current lighteval evaluation script suggested by open-r1 repository may contain some bugs, and I've made some changes.
see: https://github.com/huggingface/open-r1/issues/602#issuecomment-2812176437
```
"lighteval|aime24|0": {
      "math_pass@1:32_samples": 0.415625,
      "math_pass@1:32_samples_stderr": 0.06684182494353075,
      "extractive_match": 0.4,
      "extractive_match_stderr": 0.09097176522946843
},
"lighteval|math_500|0": {
  "extractive_match": 0.926,
  "extractive_match_stderr": 0.011718474529160475
}
```

**Comment by kiseliu** [comment]
> I got the following performance using this train recipe: The current lighteval evaluation script suggested by open-r1 repository may contain some bugs, and I've made some changes. see: [#602 (comment)](https://github.com/huggingface/open-r1/issues/602#issuecomment-2812176437)
> 
> ```
> "lighteval|aime24|0": {
>       "math_pass@1:32_samples": 0.415625,
>       "math_pass@1:32_samples_stderr": 0.06684182494353075,
>       "extractive_match": 0.4,
>       "extractive_match_stderr": 0.09097176522946843
> },
> "lighteval|math_500|0": {
>   "extractive_match": 0.926,
>   "extractive_match_stderr": 0.011718474529160475
> }
> ```

Hi, @StarLooo, I cannot see what is this model. The results are not consistent with that you mentioned in the issue #602.

**Comment by StarLooo** [comment]
> > I got the following performance using this train recipe: The current lighteval evaluation script suggested by open-r1 repository may contain some bugs, and I've made some changes. see: [#602 (comment)](https://github.com/huggingface/open-r1/issues/602#issuecomment-2812176437)
> > ```
> > "lighteval|aime24|0": {
> >       "math_pass@1:32_samples": 0.415625,
> >       "math_pass@1:32_samples_stderr": 0.06684182494353075,
> >       "extractive_match": 0.4,
> >       "extractive_match_stderr": 0.09097176522946843
> > },
> > "lighteval|math_500|0": {
> >   "extractive_match": 0.926,
> >   "extractive_match_stderr": 0.011718474529160475
> > }
> > ```
> 
> Hi, [@StarLooo](https://github.com/StarLooo), I cannot see what is this model. The results are not consistent with that you mentioned in the issue [#602](https://github.com/huggingface/open-r1/issues/602).

Hi, The results I reported here is from the `OpenR1-Qwen-7B` model I **reproduced** using the recipe and modifications you mentioned in this issue above.
While the results I reported at [#602 (comment)](https://github.com/huggingface/open-r1/issues/602#issuecomment-2812176437) is from the **official ckpts** `DeepSeek-R1-Distill-Qwen-1.5B/7B` using the latest ligheval with some modifications on the command line in open-r1's , since the incompatible problems about vllm & lighteval version.

**Comment by kiseliu** [comment]
@StarLooo I see. 

The results I put here is from the commit `9890a8d9921ecf27784a18896f3b974b357df903`.
